# Palmer Penguins 

![Penguins](https://raw.githubusercontent.com/UWASIKLK/pda-mywork/main/picture%20of%20penguins.png)

The data set is available [on GitHub](https://allisonhorst.github.io/palmerpenguins/)

---

## Introduction
***

The Palmer Penguins dataset contains various measurements of three different penguin species (Adelie, Gentoo and Chinstrap) and it commonly used for data analysis as an alternative to the Iris dataset.

These data were collected from 2007 - 2009 by [Dr. Kristen Gorman](https://www.uaf.edu/cfos/people/faculty/detail/kristen-gorman.php) 
with the [Palmer Station](https://pallter.marine.rutgers.edu/), a member of the [Long Term Ecological Research Network](https://lternet.edu/) on three islands. 

## Investigation
***


The dataset has details of 344 penguins from three species of penguin: _Adélie (152 penguins)_, _Chinstrap (68 penguins)_ and _Gentoo (124 penguins)_.

The dataset contains following variables:
 - **species:**&nbsp; penguins species (_Adélie_, _Chinstrap_ or _Gentoo_)
 - **island:**&nbsp; island name (_Dream_, _Torgersen_ or _Biscoe_)
 - **bill_length_mm:**&nbsp; bill (culmen) length in millimetres
 - **bill_depth_mm:**&nbsp; depth length in millimetres
 - **flipper_length_mm:**&nbsp; flipper length in millimetres
- **body_mass_g:**&nbsp; weights in grams
- **sex:**&nbsp; the gender of penguin

![penguin details](https://raw.githubusercontent.com/UWASIKLK/pda-mywork/main/penguin%20details.png)

### Import libraries

***

For the data analysis purpose of this dataset, I would recommend downloading following python libraries:
-	Numpy – this library is used for working with arrays and processing mathematical operations.
-	Pandas – it is used to store and manipulate structured data.
-	Matplotib.pyplot – this is used to created and customise plots.
-	Seaborn – this is another tool for creating attractive data visualisation as it has beautiful default style and colour palettes.


In [166]:
# Importing libraries:

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data download

***

The Palmer Penguins dataset was downloaded from [GitHub](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv) 

In [167]:
penguins = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv")

The data is now loaded.

## Data overview

***

First, I check the data to get an overview of what's in the dataset.

In [168]:
# View of the data set. 
# It will dispay only first 5 and last 5 rows and totals of rows and columns.

penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
3            NaN     NaN  
4         3450.0  FEMALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[344 rows x 7 columns]

In [169]:
# The function head(10) will allow me to view first 10 rows from dataset.

penguins.head(10)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       3625.0  FEMALE  
7       4675.0    MALE  
8       3475.0     NaN  
9       4250.0     NaN

In [170]:
# Similarly for tail(10) which will show the last 10 rows.

penguins.tail(10)

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
334  Gentoo  Biscoe            46.2           14.1              217.0   
335  Gentoo  Biscoe            55.1           16.0              230.0   
336  Gentoo  Biscoe            44.5           15.7              217.0   
337  Gentoo  Biscoe            48.8           16.2              222.0   
338  Gentoo  Biscoe            47.2           13.7              214.0   
339  Gentoo  Biscoe             NaN            NaN                NaN   
340  Gentoo  Biscoe            46.8           14.3              215.0   
341  Gentoo  Biscoe            50.4           15.7              222.0   
342  Gentoo  Biscoe            45.2           14.8              212.0   
343  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
334       4375.0  FEMALE  
335       5850.0    MALE  
336       4875.0     NaN  
337       6000.0    MALE  
338       4925.0  FEMALE  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE

In [171]:
# I will use the describe() function to generate statistical data like percentile, mean, str (standard deviation), etc.

penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      342.000000     342.000000         342.000000   342.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.459584       1.974793          14.061714   801.954536
min         32.100000      13.100000         172.000000  2700.000000
25%         39.225000      15.600000         190.000000  3550.000000
50%         44.450000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000

In [172]:
# The info() will provide information like number of columns, column labels, column data type, number of cells in each column (non-null values), etc.

penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


## Data clean up

***

After running the info() function I can see that some information related to bill length & depth, flipper length, body mass and sex is missing. I will investigate this further using the isnull() function which will show rows that contains null values. the rows which will return boolean **'True'** will have a null values.

In [173]:
# To find out if there are any missing values

penguins.isnull()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0      False   False           False          False              False   
1      False   False           False          False              False   
2      False   False           False          False              False   
3      False   False            True           True               True   
4      False   False           False          False              False   
..       ...     ...             ...            ...                ...   
339    False   False            True           True               True   
340    False   False           False          False              False   
341    False   False           False          False              False   
342    False   False           False          False              False   
343    False   False           False          False              False   

     body_mass_g    sex  
0          False  False  
1          False  False  
2          False  False  
3           True   True  
4          False  False  
..           ...    ...  
339         True   True  
340        False  False  
341        False  False  
342        False  False  
343        False  False  

[344 rows x 7 columns]

After running the above function, I can see that two rows have missing values. The one is for Adelie penguin and anothere for Gentoo penguin. In case of large dataset, we have an option to delete the missing rows, as they will not have a significant impact on the overall result of analysis. Or we can try to replace the missing values. I decided to go for second option and replace the missing values.

In [174]:
# This will show the details of two rows with missing values

penguins.iloc[[3,339],0:7]


species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen             NaN            NaN                NaN   
339  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g  sex  
3            NaN  NaN  
339          NaN  NaN

Now I want to find out the average values of the missing columns. First, I find out the details of Adelie's penguin from Torgersen Island. I want to see the numbers of male & female related to Adelie, because this also needs to be filled.

In [175]:
# Showing details related to Adelie, Torgersen island

a_penguins = penguins.loc[(penguins.species == 'Adelie') & (penguins.island == 'Torgersen')]
a_penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count       51.000000      51.000000          51.000000    51.000000
mean        38.950980      18.429412         191.196078  3706.372549
std          3.025318       1.339447           6.232238   445.107940
min         33.500000      15.900000         176.000000  2900.000000
25%         36.650000      17.350000         187.000000  3337.500000
50%         38.900000      18.400000         191.000000  3700.000000
75%         41.100000      19.250000         195.000000  4000.000000
max         46.000000      21.500000         210.000000  4700.000000

In [176]:
# And the gender count (how many of female and male)

a_penguins['sex'].value_counts()


sex
FEMALE    24
MALE      23
Name: count, dtype: int64

I will use the **'mean'** values for the missing Adelie row. Since there is not much difference in the numbers of male and female, I opted for a female in this case.

In [177]:
# Filling in the missing data for Adelie row (values are on two decimal places)

penguins.iloc[3,2:7] = 38.95, 18.42, 191.19, 3706.37, 'FEMALE'



In [178]:
# checking if missing data were replaced by mean values for Adelie row

penguins.iloc[[3,339],0:7]


species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen           38.95          18.42             191.19   
339  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g     sex  
3        3706.37  FEMALE  
339          NaN     NaN

The missing data for Adelie row are now filled in. I will repeat the same steps for Gentoo penguin row.

In [179]:
# Showing details related to Gentoo, Biscoe island

g_penguins = penguins.loc[(penguins.species == 'Gentoo') & (penguins.island == 'Biscoe')]
g_penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      123.000000     123.000000         123.000000   123.000000
mean        47.504878      14.982114         217.186992  5076.016260
std          3.081857       0.981220           6.484976   504.116237
min         40.900000      13.100000         203.000000  3950.000000
25%         45.300000      14.200000         212.000000  4700.000000
50%         47.300000      15.000000         216.000000  5000.000000
75%         49.550000      15.700000         221.000000  5500.000000
max         59.600000      17.300000         231.000000  6300.000000

In [180]:
# And the gender count (how many of female and male)

g_penguins['sex'].value_counts()

sex
MALE      61
FEMALE    58
Name: count, dtype: int64

I will use the mean values for missing Gentoo row and I opted for a male this time.

In [181]:
# Filling in the missing data for Gentoo row (values are on two decimal places)

penguins.iloc[339,2:7] = 47.50, 14.98, 217.18, 5076.01, 'MALE'

In [182]:
# Checking if missing data were replaced by mean values for Adelie row

penguins.iloc[[3,339],0:7]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen           38.95          18.42             191.19   
339  Gentoo     Biscoe           47.50          14.98             217.18   

     body_mass_g     sex  
3        3706.37  FEMALE  
339      5076.01    MALE

I want to check how many rows we have, where the **'sex'** column has no value.

In [185]:
# This will show how many rows have missing value in sex column.

penguins[penguins['sex'].isna()]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
8    Adelie  Torgersen            34.1           18.1              193.0   
9    Adelie  Torgersen            42.0           20.2              190.0   
10   Adelie  Torgersen            37.8           17.1              186.0   
11   Adelie  Torgersen            37.8           17.3              180.0   
47   Adelie      Dream            37.5           18.9              179.0   
246  Gentoo     Biscoe            44.5           14.3              216.0   
286  Gentoo     Biscoe            46.2           14.4              214.0   
324  Gentoo     Biscoe            47.3           13.8              216.0   
336  Gentoo     Biscoe            44.5           15.7              217.0   

     body_mass_g  sex  
8         3475.0  NaN  
9         4250.0  NaN  
10        3300.0  NaN  
11        3700.0  NaN  
47        2975.0  NaN  
246       4100.0  NaN  
286       4650.0  NaN  
324       4725.0  NaN  
336       4875.0  NaN

In general, the body mass for male penguin is higher than in female penguin. I want to find out what is the average body weight of female Adelie penguins from the island of Torgersen. I will repeat the same steps for Gentoo penguins.

In [186]:
# finding the average body mass

aa_penguins = penguins.loc[(penguins.species == 'Adelie') & (penguins.island == 'Torgersen') & (penguins.sex == 'FEMALE')]
aa_penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count       25.000000      25.000000           25.00000    25.000000
mean        37.610000      17.584800          188.40760  3408.254800
std          2.179354       0.878603            4.57813   261.179937
min         33.500000      15.900000          176.00000  2900.000000
25%         35.900000      17.000000          186.00000  3200.000000
50%         38.500000      17.500000          189.00000  3450.000000
75%         39.000000      18.000000          191.00000  3625.000000
max         41.100000      19.300000          196.00000  3800.000000

According to the above result, the average body weight of Adelie penguins from Torgersen Island is 3408.25 grams. Therefore, for row 10 I fill in 'FEMALE' and for rows 8,9 & 11 I fill in 'MALE'.

In [194]:
# Filling in missing gender data.

penguins.iloc[10,6] = 'FEMALE'
penguins.iloc[8,6] = 'MALE'
penguins.iloc[9,6] = 'MALE'
penguins.iloc[11,6] = 'MALE'


I assume, that the row 47, Adelie from Dream Island with the body mass 2975.0 grams, is 'FEMALE'. It is clear from the dataset that males are larger/heavier than females.

In [199]:
# Filling in missing gender. 

penguins.iloc[47,6] = 'FEMALE'

I will repeat the steps to find out the average body weight of Gentoo penguins from Biscoe Island.

In [201]:
gg_penguins = penguins.loc[(penguins.species == 'Gentoo') & (penguins.island == 'Biscoe') & (penguins.sex == 'FEMALE')]
gg_penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count       58.000000      58.000000          58.000000    58.000000
mean        45.563793      14.237931         212.706897  4679.741379
std          2.051247       0.540249           3.897856   281.578294
min         40.900000      13.100000         203.000000  3950.000000
25%         43.850000      13.800000         210.000000  4462.500000
50%         45.500000      14.250000         212.000000  4700.000000
75%         46.875000      14.600000         215.000000  4875.000000
max         50.500000      15.500000         222.000000  5200.000000

The average body mass as per result above is 4679.74. If the body mass is below this figure on missing rows I fill in 'FEMALE', otherwise it will be 'MALE'.

In [203]:
# Filling in missing data.

penguins.iloc[246,6] = 'FEMALE'
penguins.iloc[286,6] = 'FEMALE'
penguins.iloc[324,6] = 'MALE'
penguins.iloc[336,6] = 'MALE'

In [204]:
# Checking for null values on gender

penguins[penguins['sex'].isna()]

Empty DataFrame
Columns: [species, island, bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g, sex]
Index: []

***

### End